packages/server/src/fhir/search.ts

```diff
@@ -526,42 +549,165 @@ async function getSearchRevIncludeEntries(
+ * A search request can use cursor links if:
+ *   1. Not using offset pagination
+ *   2. Exactly one sort rule using _lastUpdated ascending
+ * @param searchRequest - The candidate search request.
+ * @returns True if the search request can use cursor links.
+ */
+function canUseCursorLinks(searchRequest: SearchRequestWithCountAndOffset): boolean {
+  return (
+    searchRequest.offset === 0 &&
+    searchRequest.sortRules?.length === 1 &&
+    searchRequest.sortRules[0].code === '_lastUpdated' &&
+    !searchRequest.sortRules[0].descending
+  );
+}
+
+/**
+ * Builds the "first", "next", and "previous" links for a search request using cursor pagination.
+ * @param searchRequest - The search request.
+ * @param entries - The search bundle entries.
+ * @param hasMore - True if there are more entries after the current page.
+ * @param result - The search bundle links.
+ */
+function buildSearchLinksWithCursor(
+  searchRequest: SearchRequestWithCountAndOffset,
+  entries: BundleEntry[],
+  hasMore: boolean | undefined,
+  result: BundleLink[]
+): void {
+  let currCursor: Cursor | undefined = undefined;
+  if (searchRequest.cursor) {
+    currCursor = parseCursor(searchRequest.cursor);
+  }
+  if (!currCursor) {
+    currCursor = {
+      version: '1',
+      instant: '',
+    };
+  }
+
+  result.push({
+    relation: 'first',
+    url: getSearchUrl({ ...searchRequest, cursor: undefined, offset: undefined }),
+  });
+
+  if (hasMore) {
     result.push({
-      relation: 'first',
-      url: getSearchUrl({ ...searchRequest, offset: 0 }),
+      relation: 'next',
+      url: getSearchUrl({
+        ...searchRequest,
+        cursor: formatCursor({ ...currCursor, instant: getMaxLastUpdated(entries) }),
+        offset: undefined,
+      }),
     });
+  }
+}
 
-    if (hasMore) {
-      result.push({
-        relation: 'next',
-        url: getSearchUrl({ ...searchRequest, offset: offset + count }),
-      });
-    }
+/**
+ * Parses a cursor string into a Cursor object.
+ * @param cursor - The cursor string.
+ * @returns The Cursor object or undefined if the cursor string is invalid.
+ */
+function parseCursor(cursor: string): Cursor | undefined {
+  const parts = splitN(cursor, '-', 2);
+  if (parts.length !== 2) {
+    return undefined;
+  }
+  return { version: parts[0], instant: parts[1] };
+}
+
+/**
+ * Formats a cursor object into a cursor string.
+ * @param cursor - The cursor object.
+ * @returns The cursor string.
+ */
+function formatCursor(cursor: Cursor): string {
+  return `${cursor.version}-${cursor.instant}`;
+}
 
-    if (offset > 0) {
-      result.push({
-        relation: 'previous',
-        url: getSearchUrl({ ...searchRequest, offset: offset - count }),
-      });
+/**
+ * Returns the maximum lastUpdated values from a set of search bundle entries.
+ * @param entries - The search bundle entries.
+ * @returns The maximum lastUpdated values.`
+ */
+function getMaxLastUpdated(entries: BundleEntry[]): string {
```

Comment:
Another edge case: if there are more than `count` records with the same `lastUpdated` that happen to line up with a given page/batch, the cursor-based `next` links will point back to the same set of records, no?

Reply:
Actually, I think this would just be a similar situation as the other edge case I called out. I don't think there's an infinite loop possibility with `>`

Reply:
Same as above, edge cases should be fixed.

✅